lib/node/runner/block_operations.go

```diff
@@ -0,0 +1,236 @@
+package runner
+
+import (
+	"time"
+
+	logging "github.com/inconshreveable/log15"
+
+	"boscoin.io/sebak/lib/block"
+	"boscoin.io/sebak/lib/common"
+	"boscoin.io/sebak/lib/errors"
+	"boscoin.io/sebak/lib/storage"
+)
+
+type SavingBlockOperations struct {
+	st  *storage.LevelDBBackend
+	log logging.Logger
+
+	saveBlock    chan block.Block
+	checkedBlock uint64 // block.Block.Height
+}
+
+func NewSavingBlockOperations(st *storage.LevelDBBackend, logger logging.Logger) *SavingBlockOperations {
+	if logger == nil {
+		logger = log
+	}
+
+	logger = logger.New(logging.Ctx{"m": "SavingBlockOperations"})
+	return &SavingBlockOperations{
+		st:           st,
+		log:          logger,
+		saveBlock:    make(chan block.Block, 10),
+		checkedBlock: common.GenesisBlockHeight,
+	}
+}
+
+func (sb *SavingBlockOperations) getNextBlock(height uint64) (nextBlock block.Block, err error) {
+	if height < sb.checkedBlock {
+		height = sb.checkedBlock
+	}
+
+	height++
+
+	nextBlock, err = block.GetBlockByHeight(sb.st, height)
+
+	return
+}
+
+func (sb *SavingBlockOperations) Check() (err error) {
+	sb.log.Debug("start to check")
+	return sb.check()
+}
+
+// continuousCheck will check the missing `BlockOperation`s continuously; if it
+// is failed, still try.
+func (sb *SavingBlockOperations) continuousCheck() {
+	for {
+		if err := sb.check(); err != nil {
+			sb.log.Error("failed to check", "error", err)
+		}
+
+		time.Sleep(5 * time.Second)
+	}
+}
+
+// check checks whether `BlockOperation`s of latest `block.Block` are saved; if
+// not it will try to catch up to the last.
+func (sb *SavingBlockOperations) check() (err error) {
+	var checked bool
+	var blk block.Block
+	for {
+		if blk, err = sb.getNextBlock(blk.Height); err != nil {
+			if err == errors.StorageRecordDoesNotExist {
+				if checked {
+					sb.log.Debug("stop checking; all the blocks are checked", "height", sb.checkedBlock)
+				}
+				err = nil
+			}
+
+			break
+		}
+
+		sb.log.Debug("check block", "block", blk)
+
+		var st *storage.LevelDBBackend
+		if st, err = sb.st.OpenBatch(); err != nil {
+			break
+		}
+
+		if err = sb.CheckByBlock(st, blk); err != nil {
+			sb.log.Error("failed to check block", "block", blk, "height", blk.Height)
+			st.Discard()
+			break
+		}
+		if err = st.Commit(); err != nil {
+			st.Discard()
+			break
+		}
+		sb.log.Debug("checked block", "block", blk)
+		sb.checkedBlock = blk.Height
+		checked = true
+	}
+
+	return
+}
+
+func (sb *SavingBlockOperations) savingBlockOperationsWorker(id int, st *storage.LevelDBBackend, blk block.Block, ops <-chan string, results chan<- error) {
+	for op := range ops {
+		results <- sb.CheckTransactionByBlock(st, blk, op)
```

Comment:
At first I saw this line, I expected that `CheckTransactionByBlock` returns some usable result,
but it returns error.
How about renaming `results` to `errors`?